compel/test/stack/spy.c

```diff
@@ -0,0 +1,368 @@
+#include <stdio.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <sys/wait.h>
+#include <sys/mman.h>
+
+#include <common/page.h>
+
+#include <compel/log.h>
+#include <compel/infect-rpc.h>
+#include <errno.h>
+
+#include "parasite.h"
+
+#define PARASITE_CMD_INC PARASITE_USER_CMDS
+#define PARASITE_CMD_DEC PARASITE_USER_CMDS + 1
+
+#define err_and_ret(msg)              \
+	do {                          \
+		fprintf(stderr, msg); \
+		return -1;            \
+	} while (0)
+
+void *saved_data = NULL;
+
+#define SAVED_DATA_MAX page_size()
+
+void cleanup_saved_data(void)
+{
+	free(saved_data);
+}
+
+static void print_vmsg(unsigned int lvl, const char *fmt, va_list parms)
+{
+	printf("\tLC%u: ", lvl);
+	vprintf(fmt, parms);
+}
+
+static void *get_parasite_rstack_start(struct parasite_ctl *ctl)
+{
+	void *rstack, *r_thread_stack, *rstack_start;
+
+	compel_get_stack(ctl, &rstack, &r_thread_stack);
+
+	rstack_start = rstack;
+	if (r_thread_stack != NULL && r_thread_stack < rstack_start)
+		rstack_start = r_thread_stack;
+
+	return rstack_start;
+}
+
+static int page_writable(struct parasite_ctl *ctl, int pid, void *page)
+{
+	FILE *maps;
+	size_t maps_line_len = 0;
+	char *maps_line = NULL;
+	char r, w;
+	int ret = 0;
+
+	if (((uintptr_t)page & (page_size() - 1)) != 0) {
+		fprintf(stderr, "Page address not aligned\n");
+		ret = -1;
+		goto done;
+	}
+
+	char victim_maps_path[6 + 11 + 5 + 1];
+	int written = snprintf(victim_maps_path, sizeof(victim_maps_path), "/proc/%d/maps", pid);
+	if (written < 0 || written >= sizeof(victim_maps_path)) {
+		fprintf(stderr, "Failed to create path string to victim's /proc/%d/maps file\n", pid);
+		ret = -1;
+		goto done;
+	}
+
+	maps = fopen(victim_maps_path, "r");
+	if (maps == NULL) {
+		perror("Can't open victim's /proc/$pid/maps");
+		ret = -1;
+		goto done;
+	}
+
+	while (getline(&maps_line, &maps_line_len, maps) != -1) {
+		unsigned long vmstart, vmend;
+		if (sscanf(maps_line, "%lx-%lx %c%c", &vmstart, &vmend, &r, &w) < 4) {
+			fprintf(stderr, "Can't parse victim's /proc/%d/maps; line: %s\n", pid, maps_line);
+			ret = -1;
+			goto free_linebuf;
+		}
+
+		if (page >= (void *)vmstart && page < (void *)vmend) {
+			if (w == 'w') {
+				if (r != 'r') {
+					fprintf(stderr, "Expecting writable memory to also be readable");
+					ret = -1;
+					goto free_linebuf;
+				}
+				ret = 1;
+			}
+			break;
+		}
+	}
+	if (errno) {
+		perror("Can't read victim's /proc/$pid/maps");
+		ret = -1;
+	}
+free_linebuf:
+	free(maps_line);
+	fclose(maps);
+done:
+	return ret;
+}
+
+static void *read_proc_mem(int pid, void *offset, size_t len)
+{
+	char victim_mem_path[6 + 11 + 4 + 1];
+	int written = snprintf(victim_mem_path, sizeof(victim_mem_path), "/proc/%d/mem", pid);
```

Comment:
Same, please do variable declaration separate from other code, and put empty line between variable declaration and other code.

Reply:
Fixed, thanks.